### PR TITLE
add dtype to batchify function

### PR DESCRIPTION
**Pull request by szhengac**
## Description ##
add dtype to batchify function

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: gluonnlp/data/batchify.py [hunk 1]
```diff
@@ -47,12 +47,19 @@ def _pad_arrs_to_max_length(arrs, pad_axis, pad_val, use_shared_mem=False):
...
     ret_shape = (len(arrs), ) + tuple(ret_shape)
     if use_shared_mem:
-        ret = mx.nd.full(shape=ret_shape, val=pad_val, ctx=mx.Context('cpu_shared', 0),
-                         dtype=arrs[0].dtype)
+        if dtype is not None:
+            ret = mx.nd.full(shape=ret_shape, val=pad_val, ctx=mx.Context('cpu_shared', 0),
+                             dtype=dtype)
+        else:
+            ret = mx.nd.full(shape=ret_shape, val=pad_val, ctx=mx.Context('cpu_shared', 0),
+                             dtype=arrs[0].dtype)
```
Comment (szha) [inline gluonnlp/data/batchify.py]:
```python
dtype = arrs[0].dtype if dtype is None else dtype
ret = mx.nd.full(shape=ret_shape, val=pad_val, ctx=mx.Context('cpu_shared', 0),
                           dtype=dtype)
```

File: tests/unittest/test_batchify.py [hunk 2]
```diff
@@ -5,18 +5,21 @@
...
-        batchify_out = batchify_fn(dat).asnumpy()
-        npy_out = np.array(dat)
-        assert_allclose(batchify_out, npy_out)
-        assert batchify_out.dtype == npy_out.dtype
+        for _dtype in [None, dtype]:
+            batchify_fn = batchify.Stack(dtype=_dtype)
+            batchify_out = batchify_fn(dat).asnumpy()
+            npy_out = np.array(dat)
+            assert_allclose(batchify_out, npy_out)
+            assert batchify_out.dtype == npy_out.dtype
```
Comment (sxjscience) [inline tests/unittest/test_batchify.py]: It has not checked the case that the `dtype` used in Stack is different from the `dtype` of the input array. We may do the following instead:
```python
for _dtype in [np.float32]:
     batchify_fn = batchify.Stack(dtype=_dtype)
     batchify_out = batchify_fn(dat).asnumpy()
     assert batchify_out.dtype == np.float32
```

File: gluonnlp/data/batchify.py [hunk 3]
```diff
@@ -66,23 +67,38 @@ def _pad_arrs_to_max_length(arrs, pad_axis, pad_val, use_shared_mem=False):
...
         else:
             return mx.nd.stack(*arrs)
     else:
         out = np.asarray(arrs)
+        dtype = out.dtype if dtype is None else dtype
         if use_shared_mem:
-            return mx.nd.array(out, ctx=mx.Context('cpu_shared', 0), dtype=out.dtype)
+            return mx.nd.array(out, ctx=mx.Context('cpu_shared', 0), dtype=dtype)
         else:
             return mx.nd.array(out, dtype=out.dtype)
```
Comment (sxjscience) [inline gluonnlp/data/batchify.py]: Should be `return mx.nd.array(out, dtype=dtype)`.

File: tests/unittest/test_batchify.py [hunk 4]
```diff
@@ -37,62 +40,69 @@ def _verify_padded_arr(padded_arr, original_arr, pad_axis, pad_val, pad_length,
...
-                                                    ret_use_mx[i][1].asnumpy())
-                                    assert(ret_use_npy[i][1].shape == (batch_size,))
-                                else:
-                                    assert ret_use_npy[i].dtype == ret_use_mx[i].dtype == dtype
-                                    assert_allclose(ret_use_npy[i].asnumpy(), ret_use_mx[i].asnumpy())
+                        for _dtype in [None, dtype]:
+                            shapes = [[2 for _ in range(ndim)] for _ in range(batch_size)]
+                            for i in range(len(shapes)):
+                                shapes[i][axis] = np.random.randint(length_min, length_max)
+                            random_data_npy = [np.random.normal(0, 1, shape).astype(dtype)
```
Comment (sxjscience) [inline tests/unittest/test_batchify.py]: Since we set the random_data_npy to have `dtype`, the check also does not test the case that the `dtypes` are different.

File: gluonnlp/data/batchify.py [hunk 5]
```diff
@@ -66,23 +67,38 @@ def _pad_arrs_to_max_length(arrs, pad_axis, pad_val, use_shared_mem=False):
...
+    r""""Base class for applying batchify function on input data.
+
+    Parameters
+    ----------
+    dtype : str or numpy.dtype, default None
+        The value type of the output. If it is set to None, the input data type is used.
+    """
+
+    def __init__(self, dtype=None):
+        self._dtype = dtype
```
Comment (sxjscience) [inline gluonnlp/data/batchify.py]: I suggest manually adding the `dtype` argument to `Stack` and `Pad`. This base class looks redundant.

Comment (szhengac) [reply gluonnlp/data/batchify.py]: This may be cleaner if we add another batchify function in the future?

Comment (sxjscience) [reply gluonnlp/data/batchify.py]: Since it has only one argument, it would be okay to add it manually to other classes. Just a single-line of code. Also, `Tuple` cannot be inherit `Batchify`.

Comment (szhengac) [reply gluonnlp/data/batchify.py]: Do we really need to have Tuple inherit Batchify? Tuple is only supposed to wrap other batchify functions.

Comment (sxjscience) [reply gluonnlp/data/batchify.py]: Not really. The main concern is that the base class has just one argument and is equivalent to a single-line of code. I think it would be better to keep the interface as simple as possible and avoid adding too many hierarchies.

Comment (szhengac) [reply gluonnlp/data/batchify.py]: Okay

Comment (szhengac) [reply gluonnlp/data/batchify.py]: But we have duplicate doc.

**Comment by mli** [comment]
Job PR-216/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-216/3/index.html

**Comment by mli** [comment]
Job PR-216/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-216/5/index.html
